### ISSUE TITLE
forest error are all NaN

### ISSUE THREAD

**Issue by newTypeGeek** [scikit-learn-contrib/forest-confidence-interval#97]
I have encountered this warning message when executing `forestci.random_forest_error(regressor, x_train, x_test)`
```
RuntimeWarning: invalid value encountered in true_divide g_eta_main = g_eta_raw / sum(g_eta_raw)
```

and the results are
```
array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan])
```

Is it something related to zero division?

**Comment by haijunli0629** [comment]
I encounter the same probrem.

**Comment by markvmiller** [comment]
I am having the same problem, as well.

**Comment by EyalSel** [comment]
I am having the same problem, as well.

**Comment by EyalSel** [comment]
fwiw my data has categorical features, which from a cursory perusal of the code I think might have played a role in this error.

**Comment by danieleongari** [comment]
I had the same issue with a `RandomForestRegressor`, but applying `StandardScaler` to the data SOLVED the issue: looks like this is a numerical issue.

**Comment by arokem** [comment]
Hey @danieleongari : do you want to add a note about this to the docs?

**Comment by danieleongari** [comment]
Sure, but I would like to first reproduce and isolate the problem to see where the non-scaled inputs make the problem raise. Do you have an idea of which line of code could go in overflow? 
ps: consider that for this I have to wait for some free time to dedicate, don't expect it soon :-)

**Comment by danieleongari** [comment]
In my testing if I use the following sample generator
```
def func(x, noise=0.1, factor=1):
    return (np.sqrt(x[0]) + x[0]*x[1] + 2*x[1] + sum(x) + noise*np.random.normal()) * factor
```
in a script e.g., 
```
n_features = 3
n_train_samples = 21
X_train = np.random.rand(n_train_samples, n_features)
y_train = np.apply_along_axis(func1d=func, axis=1, arr=X_train)
rf = RandomForestRegressor()
rf.fit(X_train, y_train)
fci.random_forest_error(
    forest=rf,
    X_train_shape=X_train.shape,
    X_test=X_train,
    inbag=None,
    calibrate=True,
    memory_constrained=False,
    memory_limit=None,
    y_output=None
)
```
when `factor>10000` I always get 
```
forest-confidence-interval\forestci\calibration.py:86: RuntimeWarning: overflow encountered in exp
  g_eta_raw = np.exp(np.dot(XX, eta)) * mask
forest-confidence-interval\forestci\calibration.py:102: RuntimeWarning: invalid value encountered in true_divide
  g_eta_main = g_eta_raw / sum(g_eta_raw)
array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan])
```
This makes clear that having large absolute values in the samples leads to overflow: `StandardScaler` by definition helps mitigating this problem, but we should look for a more stable solution which avoids the numerical ceiling on high numbers.

**Comment by el-hult** [comment]
@danieleongari the example data you have here works well with the updates I propse in #114 most of the time. depending on the random seed, it can sometimes raise a warning about numerical precision, but it does not fail with nan anymore.


<details>
<summary>Code to produce plots</summary>

```python
import forestci as fci, numpy as np
from sklearn.ensemble import RandomForestRegressor
np.random.seed(0)
def func(x, noise=0.1, factor=1): # update the factor!
    return (np.sqrt(x[0]) + x[0]*x[1] + 2*x[1] + sum(x) + noise*np.random.normal()) * factor
n_features = 3
n_train_samples = 21
X_train = np.random.rand(n_train_samples, n_features)
y_train = np.apply_along_axis(func1d=func, axis=1, arr=X_train)
forest = RandomForestRegressor(random_state=0)
forest.fit(X_train, y_train)
res=fci.random_forest_error(
    forest=forest,
    X_train_shape=X_train.shape,
    X_test=X_train,
    inbag=None,
    calibrate=True,
    memory_constrained=False,
    memory_limit=None,
    y_output=None
)
print(res)

import matplotlib.pyplot as plt
import matplotlib
matplotlib.use("Agg")
plt.errorbar(y_train, forest.predict(X_train), yerr=np.sqrt(res), fmt='o')
plt.xlabel("True value")
plt.ylabel("Prediction w/ std dev")
plt.savefig("fci_res.png")


# Code below here is a copy-and-paste from the forestci modules, but with some plotting commands intermingled
from forestci.calibration import gfit, calibrateEB, gbayes
from forestci.forestci import _centered_prediction_forest, _core_computation, _bias_correction, calc_inbag
from forestci import random_forest_error
import copy

X_train_shape = X_train.shape
X_test = X_train
memory_constrained = False
memory_limit = None
y_output = None

inbag = calc_inbag(X_train_shape[0], forest)
pred_centered = _centered_prediction_forest(forest, X_test, y_output)
n_trees = forest.n_estimators
V_IJ = _core_computation(
    X_train_shape, X_test, inbag, pred_centered, n_trees, memory_constrained, memory_limit
)
V_IJ_unbiased = _bias_correction(V_IJ, inbag, pred_centered, n_trees)


calibration_ratio = 2
n_sample = np.ceil(forest.n_estimators / calibration_ratio)
new_forest = copy.deepcopy(forest)
random_idx = np.random.permutation(len(new_forest.estimators_))[: int(n_sample)]
new_forest.estimators_ = list(np.array(new_forest.estimators_)[random_idx])
if hasattr(new_forest, "_seeds"):
    new_forest._seeds = new_forest._seeds[random_idx]

new_forest.n_estimators = int(n_sample)

results_ss = random_forest_error(
    new_forest,
    X_train_shape,
    X_test,
    calibrate=False,
    memory_constrained=memory_constrained,
    memory_limit=memory_limit,
    y_output=y_output
)

# Use this second set of variance estimates
# to estimate scale of Monte Carlo noise
sigma2_ss = np.mean((results_ss - V_IJ_unbiased) ** 2)
delta = n_sample / forest.n_estimators
sigma2 = (delta ** 2 + (1 - delta) ** 2) / (2 * (1 - delta) ** 2) * sigma2_ss

# Use Monte Carlo noise scale estimate for empirical Bayes calibration
variances = V_IJ_unbiased
assert sigma2 > 0
sigma = np.sqrt(sigma2)
eb_prior = gfit(variances, sigma)
calib_all = [gbayes(x,g_est=eb_prior,sigma=sigma) for x in variances]

fig,ax = plt.subplots()
ax.scatter(variances, calib_all)
ax.set_xlabel("Variance estimate")
ax.set_ylabel("Calibrated variance estimate")
plt.savefig("fci_calibration.png")

fig,ax = plt.subplots()
ax.hist(variances, bins='auto', alpha=0.5, label="Variance estimates (left)", density=True)
ax.hist(calib_all, bins='auto', alpha=0.5, label="Calibrated variance estimates (left)", density=True)
ax2 = ax.twinx()
ax2.plot(eb_prior[0], eb_prior[1])
ax.legend()
plt.savefig("fci_hist.png")
```
</details>

Running this example with `factor=1` using the branch of PR #114 yields plot as per below. The plot has a normalized histogram for "calibrated" and "uncalibrated" data (on left y-axis) and the empirical bayes prior (right y-axis). It shows that the empirical bayes prior actually is quite nice for this specific data.

<img src="https://github.com/scikit-learn-contrib/forest-confidence-interval/assets/11832922/93659a03-1ad2-421f-a7c4-eec8675e0ceb" width="400" />

Doing the same, but with code from `master` branch shows

<img src="https://github.com/scikit-learn-contrib/forest-confidence-interval/assets/11832922/51b89e93-47dc-4753-98ac-922f4c173d1e" width="400" />

indicating that the EB calibration is indeed quite broken.

With the large `factor=10000`, the code on `master` fails with NaN. My proposed fix complains about numerical precision, but nevertheless yields a result consistent with small factor. No NaN-values.

<img src="https://github.com/scikit-learn-contrib/forest-confidence-interval/assets/11832922/cd5a2f91-72f5-449c-901d-a55c6b738b64" width="400" />

**Comment by danieleongari** [comment]
Improved in v0.7
